Adds `teams chat get` command - Closes #2899

martinlingstuyl opened the pull request:
Closes #2899

The new `teams chat get` command can be used to get the details of a chat conversation to one or more people. An `id` can be used to get a specific chat conversation. The `name` or `participants` options can also be used, these will cause the command to try to find an existing conversation by name or members.

The command does **not** include the members or messages in the output. These can be retrieved using other cli commands.

waldekmastykarz commented:
Awesome @martinlingstuyl! We'll review it shortly

waldekmastykarz reviewed (changes requested):
Hey @martinlingstuyl, it seems that our recent refactoring is breaking your code. Could you please have another look?

martinlingstuyl commented:
Will do 👍 I thought I'd rebuild and retested after rebasing. Will check it out

martinlingstuyl commented:
Apparently I did not @waldekmastykarz 🤦‍♂️
I just fixed it.

waldekmastykarz commented:
Awesome! Thank you for such a quick turnaround 👏

waldekmastykarz commented on src/m365/teams/commands.ts:
```diff
@@ -13,6 +13,7 @@ export default {
   CHANNEL_REMOVE: `${prefix} channel remove`,
   CHANNEL_SET: `${prefix} channel set`,
   CHAT_LIST: `${prefix} chat list`,
+  CHAT_GET: `${prefix} chat get`,
```
Let's sort names alphabetically for easier lookup

waldekmastykarz commented on src/m365/teams/commands/chat/chat-get.ts:
```diff
@@ -0,0 +1,200 @@
+  public get description(): string {
+    return 'Get a chat conversations by id, participants or chat name';
+  }
+
```
Since we have some optional options in this command, let's add tracking their usage to our telemetry

waldekmastykarz commented on src/m365/teams/commands/chat/chat-get.ts:
```diff
@@ -0,0 +1,200 @@
+  }
+
+  public get description(): string {
+    return 'Get a chat conversations by id, participants or chat name';
```
`conversation` - singular, right?

waldekmastykarz commented on src/m365/teams/commands/chat/chat-get.ts:
```diff
@@ -0,0 +1,200 @@
+  }
+
+  public async commandAction(logger: Logger, args: CommandArgs, cb: () => void): Promise<void> {
+
```
We can remove this empty line

waldekmastykarz commented on src/m365/teams/commands/chat/chat-get.ts:
```diff
@@ -0,0 +1,200 @@
+    return 'Get a chat conversations by id, participants or chat name';
+  }
+
+  public async commandAction(logger: Logger, args: CommandArgs, cb: () => void): Promise<void> {
```
`commandAction` is treated as a sync method in the CLI runtime. Won't making it async cause issues?

martinlingstuyl replied:
Good question. Because it is marked `async`, the `commandAction` function will now return a `promise` instead of `void`. I have not seen any issues with the execution. Of course the cli is not awaiting the promise, but that's okay as we're using a callback function anyway. 

Strictly speaking though, I can understand if you would like it better to adhere to the interface of commandAction returning void. Everyone else is doing it as well.

I've refactored it to return void. I should refactor the teams chat message send command for this as well. Let me know how you want to organise a refactoring of an existing command.

waldekmastykarz replied:
We can use async for other methods that we call inside `commandAction`, as long as the main flow is sync and calls the callback to signal command execution completed to the CLI runtime.

martinlingstuyl replied:
https://github.com/pnp/cli-microsoft365/issues/3106

waldekmastykarz commented on src/m365/teams/commands/chat/chat-get.ts:
```diff
@@ -0,0 +1,200 @@
+      responseType: 'json'      
+    };
+
+    const chat = await request.get<Chat>(requestOptions);
```
I don't think there's a reason for us here to await the result and we could directly use `return request.get()`

waldekmastykarz commented on src/m365/teams/commands/chat/chat-get.ts:
```diff
@@ -0,0 +1,200 @@
+  }
+
+  private async getChatIdByParticipants(participantsString: string): Promise<string> {
+    const participants = participantsString.toLowerCase().replace(/\s/g, '').split(',').filter(e => e && e !== '');
```
Let's move converting comma-separated list of users to a separate method that we can share between validation and here to ensure consistency

waldekmastykarz commented on src/m365/teams/commands/chat/chat-get.ts:
```diff
@@ -0,0 +1,200 @@
+    }).join(os.EOL);
+
+    throw new Error(`Multiple chat conversations with this topic found. Please disambiguate:${os.EOL}${disambiguationText}`);
+    
```
Let's remove this empty line

waldekmastykarz commented on src/m365/teams/commands/chat/chat-get.ts:
```diff
@@ -0,0 +1,200 @@
+    const existingChats = await this.findExistingGroupChatsByTopic(name);
+
+    if (!existingChats || existingChats.length === 0) {
+      throw new Error('No chat conversation was found with this name.');
```
We use `topic` and `name` in different places. Let's decide on one and use it consistently throughout

martinlingstuyl replied:
Good point, I've changed it to `name`. Topic is now only used on the Microsoft Graph types and endpoints, but that's okay I think.

waldekmastykarz commented on src/m365/teams/commands/chat/chat-get.ts:
```diff
@@ -0,0 +1,200 @@
+  }
+  
+  private async findExistingChatsByParticipants(expectedMemberEmails: string[]): Promise<Chat[]> {
+    const chatType = expectedMemberEmails.length === 2 ? "oneOnOne" : "group";
```
Let's use single quotes consistently

waldekmastykarz commented on src/m365/teams/commands/chat/chat-get.ts:
```diff
@@ -0,0 +1,200 @@
+    const endpoint = `${this.resource}/v1.0/chats?$filter=chatType eq '${chatType}'&$expand=members&$select=id,topic,createdDateTime,members`;
+    const foundChats: Chat[] = [];
+
+    const chats = await this.getAllChats(endpoint, []);
```
We can use the `getAllItems` method from the `odata` helper class here

martinlingstuyl replied:
Ah, nice, it's been refactored. We should refactor the `chat message send` command as well to make use of this. Is that something that needs to be created as an issue? I'm seeing multiple comments here that that command will benefit from.

waldekmastykarz replied:
Good catch! Yes, let's create a separate issue for each refactoring so that we don't need to pile them up and can let others help us.

martinlingstuyl replied:
https://github.com/pnp/cli-microsoft365/issues/3107

waldekmastykarz commented on src/m365/teams/commands/chat/chat-get.ts:
```diff
@@ -0,0 +1,200 @@
+
+  private async findExistingGroupChatsByTopic(topic: string): Promise<Chat[]> {
+    const endpoint = `${this.resource}/v1.0/chats?$filter=topic eq '${encodeURIComponent(topic)}'&$expand=members&$select=id,topic,createdDateTime,chatType`;
+    const chats = await this.getAllChats(endpoint, []);
```
I don't think we need to `await` the results here and could directly return all chats

waldekmastykarz commented on docs/docs/cmd/teams/chat/chat-get.md:
```diff
@@ -0,0 +1,52 @@
+
+--8<-- "docs/cmd/_global.md"
+
+## Remarks
```
Let's add in remarks, that we're already adding the user currently signed in to CLI as a participant

waldekmastykarz reviewed (changes requested):
Great start! Let's do a few adjustments before we continue

plamber commented on src/m365/teams/commands/chat/chat-get.ts:
```diff
@@ -0,0 +1,191 @@
+  }
+
+  public get description(): string {
+    return 'Get a chat conversation by id, participants or chat name';
```
We should align the description of the documentation with that description.

martinlingstuyl replied:
Hi @plamber ,

I checked what you mean.

```powershell
m365 teams chat get -p martin@blimped.nl,adelev@blimped.nl
```

This does not work because the comma is interpreted by powershell. The string value is now interpreted as two values and concatenated with a space.

==> I'l fix it by updating the `convertCommaSeparatedListToArray()` function to be able to split by space if necessary.

```powershell
m365 teams chat get --name Let's 
```
The single quote is causing this when posting to Microsoft Graph. In OData, single quotes need to be duplicated for the query to work. 

==> I'll fix it by replacing all `'` with `''`

plamber reviewed (changes requested):
Hi @martinlingstuyl,
I did an additional review and found some minor things when testing the command.

Running the command with special characters returns invalid filter clause even if the characters are valid in the chat: e.g. m365 teams chat get --name "Let's"

On windows Powershell I am not able to call the command without quotes. e.g. 
<img width="782" alt="image" src="https://user-images.githubusercontent.com/18241305/159108774-3a3a063e-394c-4b45-a021-db1ee49812dc.png">
It works, however, if I am adding quotes or double quotes. Could you check if we can do something in the code or if we have to update the documentation?

Thank you for your help,
Patrick

martinlingstuyl commented:
Hi Patrick, good testing work! Let me see if I can reproduce (and fix) the things you're talking about.

martinlingstuyl commented:
Hi @plamber,
I fixed the issues you mentioned.

I also moved some functions that were being reused between `chat message send` and `chat get` to a shared util class.

hope to hear from you.

plamber commented:
HI @martinlingstuyl,
thank you. Will review it shortly.

Cheers

plamber commented:
Thank you. Merged manually
